### PR TITLE
Member profile page

### PR DESCRIPTION
**Pull request by reacteer**
Member profile page UI done
No api linked
https://freedcamp.com/view/2956183/tasks/40026083

Please make showProfile state to true in Sidebar in Sidebar.tsx to see profile page link on sidebar.

image

### CODE REVIEW

File: src/legacy/components/Profile.tsx [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+          'flex flex-col pb-4 rounded-md shadow-medium',
+          styles.card,
+        )}
+      >
+        <div className='bg-blue-300 h-32 rounded-t-md'></div>
+        <div className='-mt-10 px-4 flex'>
+          <div className='rounded-full border-4 border-white bg-pink-200 w-24 h-24'></div>
+          <div className='mt-11 pl-3 text-sm'>
+            <div className='px-1 text-gray-300'>@zndrson</div>
+            <div className='pt-2 text-blue-400'>0xc41923...2145</div>
```
Comment (CrazyOptimist) [inline src/legacy/components/Profile.tsx]: This is the user's wallet address, could you implement the real truncate instead of hard coding `...` ?

**Review by CrazyOptimist — CHANGES_REQUESTED** [review]
#### Could you make the screen work for various screen width?

	Below config is what Sanjan suggested, and I'm also adopting it:
	```javascript
	screens: {
		sm: ‘900px’,
		md: ‘1100px’,
		lg: ‘1300px’,
		xl: ‘1440px’,
		‘2xl’: ‘1700px’,
	 }
	 ```
     For the narrow screen width, you can get the followers section go below the "Bio".

#### Also please resolve the conflict.

**Comment by CrazyOptimist** [comment]
Worked like a charm as I tested just now.
Please resolve the conflicts, I'd love to approve!
@reacteer

**Comment by CrazyOptimist** [comment]
Don't worry about the side bar, @sanjanadhikari1 already removed it in his Header work, so you can remove it also, in the future work I mean.

File: src/features/profile/index.tsx [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+    frontLen >= strLen ||
+    backLen >= strLen ||
+    frontLen + backLen >= strLen
+  ) {
+    return str
+  } else if (backLen === 0) {
+    return str.slice(0, frontLen) + truncateStr
+  } else {
+    return str.slice(0, frontLen) + truncateStr + str.slice(strLen - backLen)
+  }
```
Comment (ngvtuan) [inline src/features/profile/index.tsx:69]: I think we don't need to use `else` here.  It just need: 

`return str.slice(0, frontLen) + truncateStr + str.slice(strLen - backLen)`

File: src/features/profile/index.tsx [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+          )}
+        >
+          {' '}
+          {follower.name}{' '}
+        </div>
+        <div className={classnames('text-gray-400 text-sm')}>161 followers</div>
+      </div>
+    )
+  }
+}
```
Comment (ngvtuan) [inline src/features/profile/index.tsx:304]: it has a lot of functions and components in this file. Can we create new components? Please refer here: 

https://github.com/pastelnetwork/pastel-electron-wallet/tree/master/src/features/receive

**Review by CrazyOptimist — APPROVED** [review]
Looks awesome!
👍 👍

**Comment by CrazyOptimist** [comment]
Prefixing has been reverted now, @reacteer

**Review by ngvtuan — COMMENTED** [review]
UI is not same with the design-1

Please also check the existing UI.

**Comment by dbizhev** [comment]
@CrazyOptimist from your past comment https://github.com/pastelnetwork/pastel-electron-wallet/pull/105#issuecomment-851140924

**Comment by dbizhev** [comment]
@reacteer the page is missing the header.

**Comment by dbizhev** [comment]
@reacteer the text "Member since May 15, 2021" should center on the left sidebar.
image

**Comment by dbizhev** [comment]
@reacteer some icons are not matching design. i.e. facebook, twitter, report, message and etc.

**Review by dbizhev — CHANGES_REQUESTED** [review]
Resolve the conflict.

**Review by dbizhev — CHANGES_REQUESTED** [review]
@reacteer all the buttons and icons are missing focus and hover states.
image

File: src/features/profile/ProfileCard.tsx [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+function truncateMiddle(
+  str: string,
+  frontLen: number,
+  backLen: number,
+  truncateStr: string,
+) {
+  if (str === null) {
+    return ''
+  }
```
Comment (br4in3x) [inline src/features/profile/ProfileCard.tsx:87]: `str` will never be `null` because it's type is `string`. This statement is not necessary

File: src/features/profile/ProfileCard.tsx [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  const strLen = str.length
+  // Setting default values
+  frontLen = ~~frontLen // will cast to integer
+  backLen = ~~backLen
```
Comment (br4in3x) [inline src/features/profile/ProfileCard.tsx:91]: Please use `Math.floor` for readability

File: src/features/profile/ProfileCard.tsx [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+function truncateMiddle(
+  str: string,
+  frontLen: number,
+  backLen: number,
+  truncateStr: string,
+) {
+  if (str === null) {
+    return ''
+  }
+  const strLen = str.length
```
Comment (br4in3x) [inline src/features/profile/ProfileCard.tsx:88]: This constant is not necessary (you just renaming `str.length` to `strLen` and it has no effect)

File: src/features/profile/ProfileCard.tsx [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  // Setting default values
+  frontLen = ~~frontLen // will cast to integer
+  backLen = ~~backLen
+  truncateStr = truncateStr || '&hellip;'
```
Comment (br4in3x) [inline src/features/profile/ProfileCard.tsx:92]: Please use default parameter for the `truncateStr` instead

File: src/legacy/components/Sidebar.tsx [hunk 8]
```diff
@@ -317,6 +317,7 @@ class Sidebar extends PureComponent<any, any> {
       exportPrivKeysModalIsOpen: false,
       exportedPrivKeys: null,
       privKeyInputValue: null,
+      showProfile: true,
```
Comment (br4in3x) [inline src/legacy/components/Sidebar.tsx:320]: Why is this needed?

File: src/legacy/containers/App.tsx [hunk 9]
```diff
@@ -3,6 +3,6 @@ import * as React from 'react'
 export default class App extends React.Component {
   render(): React.ReactElement {
     const { children } = this.props
-    return <>{children}</>
+    return <div className='overflow-y-auto overflow-x-hidden'>{children}</div>
```
Comment (br4in3x) [inline src/legacy/containers/App.tsx:6]: Why is this needed?

**Review by br4in3x — CHANGES_REQUESTED** [review]

File: src/features/profile/ProfileCard.tsx [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        <div className='font-bold text-2xl py-2'>Katy Jaison</div>
+        <div className='text-sm text-gray-71 py-1'>
+          Cosmic Perspective: Galactic Arch
+        </div>
+        <div className='text-sm text-gray-4a flex'>
+          <img
+            src={svg_location}
+            className='cursor-pointer pr-14px filter hover:contrast-200'
+          />
+          New York, US
```
Comment (dbizhev) [inline src/features/profile/ProfileCard.tsx:50]: @reacteer use dummy data at least.

File: src/features/profile/ProfileCard.tsx [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+      <div className='flex flex-col px-5'>
+        <div className='text-xs text-gray-71 pt-2'>
+          <div className='flex items-center'>
+            <StarRate />
+            <div className='pl-1 text-gray-71'>4.89 reputation</div>
+          </div>
+        </div>
+        <div className='font-bold text-2xl py-2'>Katy Jaison</div>
+        <div className='text-sm text-gray-71 py-1'>
+          Cosmic Perspective: Galactic Arch
```
Comment (dbizhev) [inline src/features/profile/ProfileCard.tsx:43]: @reacteer use dummy data at least.

File: src/features/profile/ProfileCard.tsx [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+        </div>
+      </div>
+      <div className='flex flex-col px-5'>
+        <div className='text-xs text-gray-71 pt-2'>
+          <div className='flex items-center'>
+            <StarRate />
+            <div className='pl-1 text-gray-71'>4.89 reputation</div>
+          </div>
+        </div>
+        <div className='font-bold text-2xl py-2'>Katy Jaison</div>
```
Comment (dbizhev) [inline src/features/profile/ProfileCard.tsx:41]: @reacteer use dummy data at least.

**Review by dbizhev — CHANGES_REQUESTED** [review]

File: src/features/profile/ProfileCard.tsx [hunk 13]
```diff
@@ -0,0 +1,106 @@
+import React from 'react'
```
Comment (bulatsalyakhovdev) [inline src/features/profile/ProfileCard.tsx:1]: @reacteer the profile card component is used for different pages/screens so it should be inside `/src/common/components`

**Review by bulatsalyakhovdev — CHANGES_REQUESTED** [review]

File: src/features/profile/ProfileCard.tsx (hunk 10 above)
Comment (reacteer) [reply src/features/profile/ProfileCard.tsx:50]: @dbizhev , I am not sure about format of data that will come, so I used placeholders simply.
I think we can update this when integrating the api.

Comment (dbizhev) [reply src/features/profile/ProfileCard.tsx]: @reacteer please check here: https://github.com/pastelnetwork/pastel-electron-wallet/blob/new_development/src/features/NFTMarket/NFTMarketFeed.tsx#L12-L20

File: src/common/components/Profile/ProfileCard.tsx [hunk 14]
```diff
@@ -0,0 +1,104 @@
+import React from 'react'
```
Comment (dbizhev) [inline src/common/components/Profile/ProfileCard.tsx:1]: @reacteer add the component to storybook

**Comment by dbizhev** [comment]
@reacteer what is this component?
image

File: src/common/components/Profile/ProfileCard.tsx (hunk 14 above)
Comment (reacteer) [reply src/common/components/Profile/ProfileCard.tsx:1]: @dbizhev , I am not familiar with storybook. Can we do that after merge?

File: src/features/profile/ProfileRelations.tsx [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+        >
+          Mutual (73)
+        </div>
+      </div>
+      <Followers followers={followers} />
+    </div>
+  )
+}
+
+export interface FollowersProps {
```
Comment (dbizhev) [inline src/features/profile/ProfileRelations.tsx:44]: @reacteer keep in mind to use `type` instead of `interface`

File: src/features/profile/ProfileRelations.tsx [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+  followers: FollowerProps[]
+}
+
+export interface FollowerProps {
```
Comment (dbizhev) [inline src/features/profile/ProfileRelations.tsx:48]: @reacteer keep in mind to use `type` instead of `interface`

**Review by dbizhev — CHANGES_REQUESTED** [review]

File: src/common/components/Profile/ProfileCard.tsx (hunk 14 above)
Comment (dbizhev) [reply src/common/components/Profile/ProfileCard.tsx:1]: @reacteer I'll add the profile card component to the storybook on the current branch.

File: src/common/components/Profile/StarRate.tsx [hunk 17]
```diff
@@ -0,0 +1,17 @@
+import React from 'react'
```
Comment (dbizhev) [inline src/common/components/Profile/StarRate.tsx:1]: @reacteer I think we need to rebuild the star rate component again as it's hardcoded.

File: src/features/profile/ProfileRelations.tsx [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+const ProfileRelations = (): JSX.Element => {
+  const [tab, setTab] = useState('Followers')
+  return (
+    <div className='w-full xl:w-2/5 flex flex-col flex-grow px-4 xl:pl-6 xl:pr-0'>
+      <div className='flex'>
+        <div
+          className={tab == 'Followers' ? cx_tab_active : cx_tab}
+          onClick={() => setTab('Followers')}
+        >
+          Followers (235)
```
Comment (dbizhev) [inline src/features/profile/ProfileRelations.tsx:24]: @reacteer the followers number should get from dummy data, not 235

File: src/features/profile/ProfileRelations.tsx [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+        </div>
+        <div
+          className={tab == 'Following' ? cx_tab_active : cx_tab}
+          onClick={() => setTab('Following')}
+        >
+          Following (162)
```
Comment (dbizhev) [inline src/features/profile/ProfileRelations.tsx:30]: @reacteer same for following number

File: src/features/profile/ProfileRelations.tsx [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+        </div>
+        <div
+          className={tab == 'Mutual' ? cx_tab_active : cx_tab}
+          onClick={() => setTab('Mutual')}
+        >
+          Mutual (73)
```
Comment (dbizhev) [inline src/features/profile/ProfileRelations.tsx:36]: @reacteer same for mutual number

File: src/common/components/Profile/StarRate.tsx (hunk 17 above)
Comment (reacteer) [reply src/common/components/Profile/StarRate.tsx:1]: @dbizhev , I will ask design team for empty star icon.

**Comment by ArthurVogelsang** [comment]
@reacteer Please check design again, updated design for header, you can use `Multi toggle switch` on new_development branch
image

**Comment by ArthurVogelsang** [comment]
@reacteer 
image

**Comment by ArthurVogelsang** [comment]
@reacteer 
image

**Comment by ArthurVogelsang** [comment]
@reacteer 
image

**Comment by ArthurVogelsang** [comment]
@reacteer 
image

File: src/features/profile/memberProfile/index.tsx [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+import ProfileCard from '../components/ProfileCard'
+import ProfileRelations from '../components/ProfileRelations'
+import ProfileTabs from '../components/ProfileTabs'
+import ProfileGeneral from '../components/ProfileGeneral'
+
+const Wrapper = styled.div`
+  height: calculate(100vh - 40px);
+  overflow-x: auto;
+`
+
```
Comment (ArthurVogelsang) [inline src/features/profile/memberProfile/index.tsx:12]: @reacteer Please remove styled-components, we don't use it anymore

**Review by ArthurVogelsang — CHANGES_REQUESTED** [review]

**Comment by ArthurVogelsang** [comment]
@reacteer please remove unused file src/features/profile/Profile.module.css

**Comment by muhadovic** [comment]
Please follow the icon style. (file name: ico-***.svg). @reacteer

File: package.json [hunk 22]
```diff
@@ -67,14 +67,16 @@
     "postcss-loader": "4.2.0",
     "progress-stream": "2.0.0",
     "qrcode.react": "1.0.0",
+    "rc-rate": "2.9.1",
```
Comment (ngvtuan) [inline package.json:70]: Do we need to use this package? I don't see it used in the code. If we don't need to use it. Please remove

File: package.json [hunk 23]
```diff
@@ -67,14 +67,16 @@
...
     "react": "17.0.1",
     "react-accessible-accordion": "3.3.4",
     "react-circular-progressbar": "2.0.4",
     "react-dom": "17.0.1",
     "react-hot-loader": "4.13.0",
     "react-modal": "3.11.1",
-    "react-range": "1.8.8",
     "react-number-format": "4.6.0",
+    "react-range": "1.8.8",
+    "react-ratings-declarative": "3.4.1",
```
Comment (ngvtuan) [inline package.json:79]: Can we use `rc-rate` instead of `react-ratings-declarative`? It seems `react-ratings-declarative` module hasn't updated since 3 years ago.

File: src/features/profile/memberProfile/index.tsx [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+        <div className='wrapper flex py-8 px-60px'>
+          <div className='flex w-full'>
+            <div className='flex flex-col items-center lg:justify-between'>
+              <ProfileCard {...profile_data} isMyProfile={true} />
+              <div className='text-gray-400 text-sm mt-24 lg:mt-0'>
+                Member Since May 15, 2021
+              </div>
+            </div>
+            <div className='flex flex-col flex-grow pl-8'>
+              {/* <ProfileTabs /> */}
```
Comment (ngvtuan) [inline src/features/profile/memberProfile/index.tsx:67]: Please remove the un-use code.

File: tailwind.config.js [hunk 25]
```diff
@@ -152,20 +187,34 @@ module.exports = {
...
+        '5px': '0.3125rem',
+      },
       lineHeight: {
         '11px': '0.6875rem',
       },
       container: {
         center: true,
       },
+      maxHeight: {
+        700: '700px',
```
Comment (ngvtuan) [inline tailwind.config.js:210]: Please use `rem`

**Review by ngvtuan — COMMENTED** [review]
ui issues 1

ui issues 2
`2` Missing the tooltip

File: package.json (hunk 23 above)
Comment (reacteer) [reply package.json:79]: @ngvtuan , I used rc-rate but it is pretty much different with our design. If react-ratings-declarative doesn't meet our requirements, I will change it down the road.

File: src/legacy/Routes.tsx [hunk 26]
```diff
@@ -8,6 +8,7 @@ import cstyles from './components/Common.module.css'
 import routes from './constants/routes.json'
 import App from './containers/App'
 import Dashboard from './components/Dashboard'
+import Profile from '../features/profile/memberProfile'
```
[the remark this answers is not in the document]
Comment (reacteer) [reply src/legacy/Routes.tsx:11]: @ngvtuan , I will put myProfile subfolder in this profile folder for my profile page, so I made it in memberProfile subfolder.
Let me know your thought.

File: src/index.css [hunk 27]
```diff
@@ -159,3 +144,24 @@ input[type='number']::-webkit-outer-spin-button {
 .focus-visible:focus-visible {
   outline: 1px solid #3f9af7;
 }
+
+::-webkit-scrollbar {
+  width: 3px;
```
[the remark this answers is not in the document]
Comment (reacteer) [reply src/index.css:149]: @ngvtuan , I added this style to customize scrollbar app-wise. Please let me know if you can do this better using Tailwind.

File: src/common/components/Icon/Icon.styles.ts [hunk 28]
```diff
@@ -0,0 +1,30 @@
+import styled, { css } from 'styled-components'
+
```
Comment (muhadovic) [inline src/common/components/Icon/Icon.styles.ts:2]: @reacteer  Please don't use 'styled-components' anymore.

File: src/features/profile/components/ProfileCard.tsx [hunk 29]
```diff
@@ -0,0 +1,125 @@
...
+import svg_avatar from '../../../common/assets/images/avatar.svg'
+import svg_copy from '../../../common/assets/icons/ico-copy2.svg'
+import svg_plus from '../../../common/assets/icons/ico-plus-white.svg'
+import svg_envelope from '../../../common/assets/icons/ico-envelope.svg'
+import svg_location from '../../../common/assets/icons/ico-location.svg'
+import svg_flag from '../../../common/assets/icons/ico-flag.svg'
+import svg_facebook from '../../../common/assets/icons/ico-facebook.svg'
+import svg_twitter from '../../../common/assets/icons/ico-twitter.svg'
+
+export type TProfileCardProps = {
```
Comment (muhadovic) [inline src/features/profile/components/ProfileCard.tsx:13]: Please use the "TProfileCard". Not need "Props"

File: src/features/profile/components/ProfileFollowers.tsx [hunk 30]
```diff
@@ -0,0 +1,381 @@
...
+import svg_oval_2 from '../../../common/assets/images/avatars/oval-2.svg'
+import svg_oval_3 from '../../../common/assets/images/avatars/oval-3.svg'
+import svg_oval_4 from '../../../common/assets/images/avatars/oval-4.svg'
+import svg_oval_5 from '../../../common/assets/images/avatars/oval-5.svg'
+import svg_oval_6 from '../../../common/assets/images/avatars/oval-6.svg'
+import svg_oval_7 from '../../../common/assets/images/avatars/oval-7.svg'
+
+export type TFollowersProps = TFollowerProps[]
+
+export type TFollowerProps = {
```
Comment (muhadovic) [inline src/features/profile/components/ProfileFollowers.tsx:13]: TFollower

File: src/features/profile/components/ProfileGeneral.tsx [hunk 31]
```diff
@@ -0,0 +1,105 @@
+import React from 'react'
+import StarRate from './StarRate'
+
+export type TProfileGeneralProps = {
```
Comment (muhadovic) [inline src/features/profile/components/ProfileGeneral.tsx:4]: Please remove "props"

File: src/features/profile/components/StarRate.tsx [hunk 32]
```diff
@@ -0,0 +1,59 @@
+import React from 'react'
+import Ratings from 'react-ratings-declarative'
+
+type TStarRateProps = {
```
Comment (muhadovic) [inline src/features/profile/components/StarRate.tsx:4]: Please remove "props"

**Review by muhadovic — CHANGES_REQUESTED** [review]

**Comment by ArthurVogelsang** [comment]
@reacteer Please fix `warnings` on console
image

File: src/features/profile/components/ProfileTabs.tsx [hunk 33]
```diff
@@ -0,0 +1,28 @@
...
+        Board
+        <div className='text-10px w-17px h-15px rounded-full bg-gray-a0 text-white flex justify-center items-center ml-2'>
+          {boardCount}
+        </div>
+      </div>
+    </div>
+  )
+}
+
+export default ProfileTabs
```
Comment (ArthurVogelsang) [inline src/features/profile/components/ProfileTabs.tsx:28]: @reacteer Please remove unused file

File: src/features/profile/components/ProfileGeneral.tsx [hunk 34]
```diff
@@ -0,0 +1,105 @@
...
+          <div className='flex-grow text-gray-4a font-medium'>
+            {totalItemsSold}
+          </div>
+        </div>
+        <div className='flex pt-4'>
+          <div className='w-190px text-gray-71'>Top category persentage</div>
+          <div className='flex-grow text-gray-4a font-medium'>
+            {topCategoryPercentage}
+          </div>
+        </div>
```
Comment (ArthurVogelsang) [inline src/features/profile/components/ProfileGeneral.tsx:90]: @reacteer this part is repeated on this file, I think we can make it as component & use array.map, if so, reduce code & good quality

**Review by ArthurVogelsang — CHANGES_REQUESTED** [review]

**Comment by reacteer** [comment]
> @reacteer Please fix `warnings` on console
> image

@ArthurVogelsang , this warning comes from Star Rating library. Do you think we have to change library to get rid of this?

**Comment by ArthurVogelsang** [comment]
@reacteer Maybe we can use another library. anyway, it's not good to show warnings on console for good quality

**Review by ArthurVogelsang — APPROVED** [review]
